File: frontend/src/utils/patientAge.js
const MIN_BIRTH_YEAR = 1900;
const MAX_AGE = 150;

export function getCurrentYear() {
  return new Date().getFullYear();
}

export function deriveAgeFromBirthYear(birthYear) {
  const year = Number(birthYear);
  if (!Number.isFinite(year)) {
    return null;
  }
  const age = getCurrentYear() - year;
  if (age < 0 || age > MAX_AGE) {
    return null;
  }
  return age;
}

export function validateBirthYearInput(value) {
  const year = Number(value);
  const currentYear = getCurrentYear();
  if (!Number.isFinite(year) || !Number.isInteger(year)) {
    throw new Error("Please enter a valid birth year.");
  }
  if (year < MIN_BIRTH_YEAR) {
    throw new Error(`Birth year must be ${MIN_BIRTH_YEAR} or later.`);
  }
  if (year > currentYear) {
    throw new Error("Birth year cannot be in the future.");
  }
  return year;
}

export function getPatientBirthYear(patient) {
  if (patient?.birthYear != null) {
    const year = Number(patient.birthYear);
    if (Number.isFinite(year)) {
      return year;
    }
  }
  if (patient?.age != null) {
    const age = Number(patient.age);
    if (Number.isFinite(age) && age >= 0 && age <= MAX_AGE) {
      return getCurrentYear() - Math.round(age);
    }
  }
  return null;
}

export function getPatientAge(patient) {
  const fromBirthYear = deriveAgeFromBirthYear(getPatientBirthYear(patient));
  if (fromBirthYear != null) {
    return fromBirthYear;
  }
  const age = Number(patient?.age);
  if (Number.isFinite(age) && age >= 0 && age <= MAX_AGE) {
    return Math.round(age);
  }
  return null;
}

export function formatPatientAge(patient) {
  const age = getPatientAge(patient);
  return age != null ? `${age} yrs` : "—";
}

export function patientAgeApiPayload(patient) {
  if (!patient) {
    return { patient_birth_year: null, patient_age: null };
  }
  const birthYear = getPatientBirthYear(patient);
  return {
    patient_birth_year: birthYear,
    patient_age: getPatientAge(patient),
  };
}

File: backend/app/services/patient_age.py
"""Birth year validation and approximate age derivation for patient context."""

from __future__ import annotations

from datetime import datetime

MIN_BIRTH_YEAR = 1900
MAX_AGE = 150


def current_year() -> int:
    return datetime.now().year


def validate_birth_year(birth_year: int | None) -> int | None:
    if birth_year is None:
        return None
    year = int(birth_year)
    current = current_year()
    if year < MIN_BIRTH_YEAR or year > current:
        raise ValueError(
            f"Birth year must be between {MIN_BIRTH_YEAR} and {current}."
        )
    return year


def derive_age_from_birth_year(birth_year: int | None) -> int | None:
    if birth_year is None:
        return None
    age = current_year() - int(birth_year)
    if 0 <= age <= MAX_AGE:
        return age
    return None


def resolve_patient_age(
    *,
    patient_age: int | None = None,
    patient_birth_year: int | None = None,
) -> int | None:
    derived = derive_age_from_birth_year(patient_birth_year)
    if derived is not None:
        return derived
    if patient_age is None:
        return None
    age = int(patient_age)
    if 0 <= age <= MAX_AGE:
        return age
    return None
